### PR TITLE
Fix a pile of bugs

### PR DESCRIPTION
**Pull request by WhyNotHugo**
This fixes a bunch of bugs that we didn't get to iron out before. It's still currently failing to completely run the test experiment.

This PR is a work-in-progress, as it still lacks proper unit tests.

### CODE REVIEW

File: petri_dish/app.py [hunk 1]
```diff
@@ -75,6 +77,17 @@ def get_all_subjects(self):
...
         grouped_subjects = self.subject_sink.read()
+        if grouped_subjects is None:
+            grouped_subjects = pandas.DataFrame(
+                columns=[
+                    self.index_column_name,
+                    # XXX: These columns need to be added to the DF even if
+                    # they're absent.
+                    self.GROUP_COLUMN_NAME,
+                    self.STAGE_COLUMN_NAME,
+                    self.JOINED_COLUMN_NAME,
```
Comment (nicosandller) [inline petri_dish/app.py]: Can you remind me what is the purpose of this column?

**Review by nicosandller — APPROVED** [review]
approved!

File: petri_dish/app.py (hunk 1 above)
Comment (WhyNotHugo) [reply petri_dish/app.py]: It's the date the subject joined the experiment. I'll make a note to document it.

Comment (nicosandller) [reply petri_dish/app.py]: brilliant, thank you

**Review by nicosandller — APPROVED** [review]

**Comment by nicosandller** [comment]
@WhyNotHugo meeeeergeee

**Review by TylerHendrickson — APPROVED** [review]
Deferring to the previous reviewer. Approved!

**Pull request merged**